server/src/test/java/keywhiz/service/providers/ClientAuthFactoryTest.java

```diff
@@ -193,37 +206,71 @@ public void clientWhenPrincipalAbsentThrows() {
     assertThat(factory.provide(request, httpServletRequest)).isEqualTo(client);
   }
 
-  @Test(expected = NotAuthorizedException.class)
-  public void rejectsXfcc_requesterAuthMissing() {
+  @Test public void returnsClientWhenClientPresent_fromXfccHeader_customSpiffeIdHeader() {
     when(httpServletRequest.getLocalPort()).thenReturn(xfccAllowedPort);
     when(request.getRequestHeader(ClientAuthFactory.XFCC_HEADER_NAME)).thenReturn(
         List.of(xfccHeader));
-    when(securityContext.getUserPrincipal()).thenReturn(null);
+    when(securityContext.getUserPrincipal()).thenReturn(xfccPrincipal);
 
-    factory.provide(request, httpServletRequest);
+    when(xfccSourceConfig.callerSpiffeIdHeader()).thenReturn(callerSpiffeHeaderName);
+    when(request.getRequestHeader(callerSpiffeHeaderName)).thenReturn(List.of(clientSpiffe));
+
+    assertThat(factory.provide(request, httpServletRequest)).isEqualTo(client);
   }
 
-  @Test(expected = NotAuthorizedException.class)
-  public void rejectsXfcc_requesterNameNotAllowed() {
+  @Test
+  public void returnsClientWhenClientPresent_fromXfccHeader_customSpiffeIdHeader_invalidSpiffe() {
     when(httpServletRequest.getLocalPort()).thenReturn(xfccAllowedPort);
     when(request.getRequestHeader(ClientAuthFactory.XFCC_HEADER_NAME)).thenReturn(
         List.of(xfccHeader));
     when(securityContext.getUserPrincipal()).thenReturn(xfccPrincipal);
 
-    when(xfccSourceConfig.allowedClientNames()).thenReturn(List.of());
-    when(xfccSourceConfig.allowedSpiffeIds()).thenReturn(List.of(xfccSpiffe));
+    when(xfccSourceConfig.callerSpiffeIdHeader()).thenReturn(callerSpiffeHeaderName);
+    when(request.getRequestHeader(callerSpiffeHeaderName)).thenReturn(List.of("not-spiffe-id"));
+
+    assertThat(factory.provide(request, httpServletRequest)).isEqualTo(client);
```

Comment:
Why is this test passing? `not-spiffe-id` in the caller-identification header should be unable to identify a client.

(In this case, I suspect the check is somehow falling back to the XFCC header, but we should definitely have a test that an invalid SPIFFE ID in the caller-identification header can't retrieve a client if the caller-identification header is the only one provided.)

Reply:
I changed the behavior so now the request would be rejected if the URI doesn't follow SPIFFE schema.

Reply:
Ah, I see how this works now :) (I hadn't realized this was falling back to the client identity in the xfccHeader.)